okhttp/src/test/java/io/grpc/okhttp/ExceptionHandlingFrameWriterTest.java

```diff
@@ -74,4 +89,64 @@ public void nullMessage() {
     assertThat(e.getMessage()).isNull();
     assertThat(getLogLevel(e)).isEqualTo(Level.INFO);
   }
+
+  private static Buffer createMessageFrame(String message) {
+    return createMessageFrame(message.getBytes(UTF_8));
+  }
+
+  private static Buffer createMessageFrame(byte[] message) {
+    Buffer buffer = new Buffer();
+    buffer.writeByte(0 /* UNCOMPRESSED */);
+    buffer.writeInt(message.length);
+    buffer.write(message);
+    return buffer;
+  }
+
+  /**
+   * Test logging is functioning correctly for client sent Http/2 frames. Not intended to test
+   * actual frame content being logged.
+   */
+  @Test
+  public void testFrameLogger() {
+    when(log.isLoggable(any(Level.class))).thenReturn(true);
+
+    exceptionHandlingFrameWriter.headers(100, new ArrayList<Header>());
+    verify(log).log(any(Level.class), startsWith(Direction.OUTBOUND + " HEADERS: streamId=" + 100));
+
+    final String message = "Hello Server";
+    Buffer buffer = createMessageFrame(message);
+    exceptionHandlingFrameWriter.data(false, 100, buffer, (int) buffer.size());
+    verify(log).log(any(Level.class), startsWith(Direction.OUTBOUND + " DATA: streamId=" + 100));
+
+    // At most 64 bytes of data frame will be logged.
+    exceptionHandlingFrameWriter
+        .data(false, 3, createMessageFrame(new String(new char[1000])), 1000);
+    ArgumentCaptor<String> logCaptor = ArgumentCaptor.forClass(String.class);
+    verify(log, atLeastOnce()).log(any(Level.class), logCaptor.capture());
+    String data = logCaptor.getValue();
+    assertTrue(data.endsWith("..."));
```

Comment:
Optional: you can use `Truth.assertThat(data).endsWith("...")` which prints better output if test fails. But it's OK.